src/Gui/DlgKeyboardImp.cpp

```diff
@@ -71,6 +77,231 @@ DlgCustomKeyboardImp::DlgCustomKeyboardImp( QWidget* parent  )
+
+    QObject::connect(combo, QOverload<int>::of(&QComboBox::activated), [=](int index) {
+        QByteArray current;
+        if (auto item = commandTreeWidget->currentItem())
+            current = item->data(1, Qt::UserRole).toByteArray();
+
+        commandTreeWidget->clear();
+        CommandManager & cCmdMgr = Application::Instance->commandManager();
+        QString group = combo->itemData(index, Qt::UserRole).toString();
+        auto cmds = group == QStringLiteral("All") ? cCmdMgr.getAllCommands()
+                                                   : cCmdMgr.getGroupCommands(group.toLatin1());
+        QTreeWidgetItem *currentItem = nullptr;
+        for (const Command *cmd : cmds) {
+            QTreeWidgetItem* item = new QTreeWidgetItem(commandTreeWidget);
+            if (dynamic_cast<const MacroCommand*>(cmd)) {
+                item->setText(1, QString::fromUtf8(cmd->getMenuText()));
+                item->setToolTip(1, QString::fromUtf8(cmd->getToolTipText()));
+            } else {
+                item->setText(1, qApp->translate(cmd->className(), cmd->getMenuText()));
+                item->setToolTip(1, qApp->translate(cmd->className(), cmd->getToolTipText()));
+            }
+            item->setData(1, Qt::UserRole, QByteArray(cmd->getName()));
+            item->setSizeHint(0, QSize(32, 32));
+            if (auto pixmap = cmd->getPixmap())
+                item->setIcon(0, BitmapFactory().iconFromTheme(pixmap));
+            item->setText(2, cmd->getShortcut());
+            if (auto accel = cmd->getAccel())
+                item->setText(3, QKeySequence(QString::fromLatin1(accel)).toString());
+
+            if (current == cmd->getName())
+                currentItem = item;
+        }
+        if (currentItem)
+            commandTreeWidget->setCurrentItem(currentItem);
+        commandTreeWidget->resizeColumnToContents(2);
+        commandTreeWidget->resizeColumnToContents(3);
+    });
+
+    QObject::connect(ShortcutManager::instance(), &ShortcutManager::shortcutChanged,
+                     combo, [combo]() { combo->activated(combo->currentIndex()); });
+
+    populateCommandGroups(combo);
+
+    return Application::Instance->commandManager().signalChanged.connect([combo](){
+        if (combo) {
+            populateCommandGroups(combo);
+            combo->activated(combo->currentIndex());
+        }
+    });
+}
+
+void DlgCustomKeyboardImp::initPriorityList(QTreeWidget *priorityList,
+                                            QAbstractButton *buttonUp,
+                                            QAbstractButton *buttonDown)
+{
+    QStringList labels;
+    labels << tr("Name") << tr("Title");
+    priorityList->setHeaderLabels(labels);
+    priorityList->header()->hide();
+    priorityList->header()->setSectionResizeMode(0, QHeaderView::ResizeToContents);
+    priorityList->header()->setSectionResizeMode(1, QHeaderView::ResizeToContents);
+
+    auto updatePriorityList = [priorityList](bool up) {

```

Comment:
Using a boolean here feels pretty kludgy. I can live with it here since it's in such an isolated use-case, but I think it would be better to use a more explicit UP/DOWN specifier.